File: apps/desktop-tauri/src-tauri/src/bridge/types/views/operations.rs
//! Operator-surfaces view types per
//! docs/superpowers/specs/2026-05-20-desktop-operator-surfaces-design.md
//! "Operations Snapshot Type" (line ~799). Stubs only — the panels in their
//! own PRs (#276/#277/#278/#281/#283/#284/#285/#286/#288) build and populate
//! these structs.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct DesktopOperationsSnapshot {
    pub fetched_at: String,
    pub agent_did: Option<String>,
    pub liveness: Option<RuntimeLivenessView>,
    pub liveness_unavailable_reason: Option<String>,
    pub backgrounded_tools: Vec<BackgroundedToolView>,
    pub stuck_diagnostics: Vec<StuckWorkDiagnosticView>,
    pub lineage: Option<SubagentTreeView>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct RuntimeLivenessView {
    pub expired_processing_count: i64,
    pub requests: Vec<ActiveRequestView>,
    pub active_tool_calls: Vec<ActiveToolCallView>,
    pub active_native_executors_available: bool,
    pub active_native_executors: Vec<NativeExecutorStatusView>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ActiveRequestView {
    pub request_id: String,
    pub claimed_at: Option<String>,
    pub deadline: Option<String>,
    pub deadline_expired: bool,
    pub deadline_age_ms: Option<i64>,
    pub last_progress_age_ms: i64,
    pub subagent_depth: i64,
    pub caused_by_parent_request_id: Option<String>,
    pub caused_by_trigger_kind: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ActiveToolCallView {
    pub request_id: String,
    pub tool_call_id: String,
    pub tool_name: String,
    pub started_at: Option<String>,
    pub deadline_at: Option<String>,
    pub await_mode: Option<String>,
    pub running_age_ms: i64,
    pub deadline_expired: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct NativeExecutorStatusView {
    pub id: i64,
    pub pid: u32,
    pub argv0: String,
    pub tool_name: Option<String>,
    pub started_at: String,
    pub age_ms: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct BackgroundedToolView {
    pub request_id: String,
    pub tool_call_id: String,
    pub tool_name: String,
    pub lifecycle_state: Option<String>,
    pub status: Option<String>,
    pub started_at: Option<String>,
    pub age_ms: Option<i64>,
    pub deadline_at: Option<String>,
    pub deadline_expired: bool,
    pub await_mode: Option<String>,
    pub cancel_policy: Option<String>,
    pub child_request_id: Option<String>,
    pub native_executor: Option<NativeExecutorStatusView>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct StuckWorkDiagnosticView {
    pub request_id: String,
    pub session_id: Option<String>,
    pub severity: String,
    pub reason: String,
    pub deadline_age_ms: Option<i64>,
    pub last_progress_age_ms: Option<i64>,
    pub tool_call_id: Option<String>,
    pub tool_name: Option<String>,
    pub stuck_since: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SubagentTreeView {
    pub root_request_id: String,
    pub nodes: Vec<SubagentNodeView>,
    pub edges: Vec<SubagentEdgeView>,
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SubagentNodeView {
    pub request_id: String,
    pub session_id: Option<String>,
    pub agent_did: Option<String>,
    pub behavior_id: Option<String>,
    pub lifecycle_state: Option<String>,
    pub status: Option<String>,
    pub subagent_depth: Option<i64>,
    pub caused_by_parent_request_id: Option<String>,
    pub caused_by_parent_tool_call_id: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SubagentEdgeView {
    pub parent_request_id: String,
    pub child_request_id: String,
    pub parent_tool_call_id: Option<String>,
    pub tool_name: Option<String>,
    pub await_mode: Option<String>,
    pub cancel_policy: Option<String>,
    pub lifecycle_state: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CascadeCancelPreview {
    pub root_request_id: String,
    pub preview_signature: String,
    pub root_state: Option<String>,
    pub will_interrupt: Vec<CascadeAffectedRequest>,
    pub will_detach: Vec<CascadeAffectedRequest>,
    pub already_terminal: Vec<CascadeAffectedRequest>,
    pub unknown_policy: Vec<CascadeAffectedRequest>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CascadeAffectedRequest {
    pub request_id: String,
    pub session_id: Option<String>,
    pub behavior_id: Option<String>,
    pub lifecycle_state: Option<String>,
    pub parent_request_id: Option<String>,
    pub parent_tool_call_id: Option<String>,
    pub tool_name: Option<String>,
    pub await_mode: Option<String>,
    pub cancel_policy: Option<String>,
}

/// Result envelope for `desktop_interrupt_request`. Field semantics are
/// normative per the design spec line 922–942:
/// - `accepted = true` iff the bridge latched (or confirmed already-latched)
///   `interrupt_requested_at` for `request_id`.
/// - `already_interrupted = true` iff the field was non-null prior to the
///   call; `accepted` is still `true` in that case.
/// - `stale_preview = true` is mutually exclusive with `accepted = true`.
///   On signature mismatch the bridge returns `accepted: false`,
///   `stale_preview: true`, and a fresh `preview` for the UI to redraw.
/// - `interrupt_requested_at` is the canonical timestamp the bridge observed
///   on the document after the call. Null only on a non-already-interrupted
///   failure.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct InterruptRequestResult {
    pub request_id: String,
    pub accepted: bool,
    pub interrupt_requested_at: Option<String>,
    pub already_interrupted: bool,
    pub stale_preview: bool,
    pub preview: Option<CascadeCancelPreview>,
}

File: apps/gents-desktop/src/lib/types/operations.ts
// 1:1 mirror of bridge/types/views/operations.rs. Keep these in sync.
// Panels in their own PRs import from "../lib/types" (re-exported below).

export type DesktopOperationsSnapshot = {
  fetchedAt: string;
  agentDid?: string | null;
  liveness?: RuntimeLivenessView | null;
  livenessUnavailableReason?: string | null;
  backgroundedTools: BackgroundedToolView[];
  stuckDiagnostics: StuckWorkDiagnosticView[];
  lineage?: SubagentTreeView | null;
};

export type RuntimeLivenessView = {
  expiredProcessingCount: number;
  requests: ActiveRequestView[];
  activeToolCalls: ActiveToolCallView[];
  activeNativeExecutorsAvailable: boolean;
  activeNativeExecutors: NativeExecutorStatusView[];
};

export type ActiveRequestView = {
  requestId: string;
  claimedAt?: string | null;
  deadline?: string | null;
  deadlineExpired: boolean;
  deadlineAgeMs?: number | null;
  lastProgressAgeMs: number;
  subagentDepth: number;
  causedByParentRequestId?: string | null;
  causedByTriggerKind?: string | null;
};

export type ActiveToolCallView = {
  requestId: string;
  toolCallId: string;
  toolName: string;
  startedAt?: string | null;
  deadlineAt?: string | null;
  awaitMode?: string | null;
  runningAgeMs: number;
  deadlineExpired: boolean;
};

export type NativeExecutorStatusView = {
  id: number;
  pid: number;
  argv0: string;
  toolName?: string | null;
  startedAt: string;
  ageMs: number;
};

export type BackgroundedToolView = {
  requestId: string;
  toolCallId: string;
  toolName: string;
  lifecycleState?: string | null;
  status?: string | null;
  startedAt?: string | null;
  ageMs?: number | null;
  deadlineAt?: string | null;
  deadlineExpired: boolean;
  awaitMode?: string | null;
  cancelPolicy?: string | null;
  childRequestId?: string | null;
  stuckSince?: string | null;
  cancelPendingRemoteAck: boolean;
  nativeExecutor?: NativeExecutorStatusView | null;
};

export type StuckWorkDiagnosticView = {
  requestId: string;
  sessionId?: string | null;
  severity: "warning" | "critical";
  reason: "expiredProcessing" | "expiredTool" | "stuckTool" | "pendingRemoteCancelAck";
  deadlineAgeMs?: number | null;
  lastProgressAgeMs?: number | null;
  toolCallId?: string | null;
  toolName?: string | null;
  stuckSince?: string | null;
};

export type SubagentTreeView = {
  rootRequestId: string;
  nodes: SubagentNodeView[];
  edges: SubagentEdgeView[];
  truncated: boolean;
  /** Deployments that could not be queried; their branches may be missing. */
  partialErrors?: string[];
};

export type SubagentNodeView = {
  requestId: string;
  /** Peer label the row was resolved from; absent = the local node. */
  resolvedVia?: string | null;
  sessionId?: string | null;
  agentDid?: string | null;
  behaviorId?: string | null;
  lifecycleState?: string | null;
  status?: string | null;
  subagentDepth?: number | null;
  causedByParentRequestId?: string | null;
  causedByParentToolCallId?: string | null;
  backendId?: string | null;
};

export type SubagentEdgeView = {
  parentRequestId: string;
  childRequestId: string;
  parentToolCallId?: string | null;
  toolName?: string | null;
  awaitMode?: "foreground" | "background" | string | null;
  cancelPolicy?: "cascade" | "detach" | string | null;
  lifecycleState?: string | null;
};

export type CascadeCancelPreview = {
  rootRequestId: string;
  previewSignature: string;
  rootState?: string | null;
  willInterrupt: CascadeAffectedRequest[];
  willDetach: CascadeAffectedRequest[];
  alreadyTerminal: CascadeAffectedRequest[];
  unknownPolicy: CascadeAffectedRequest[];
};

export type CascadeAffectedRequest = {
  requestId: string;
  sessionId?: string | null;
  behaviorId?: string | null;
  lifecycleState?: string | null;
  parentRequestId?: string | null;
  parentToolCallId?: string | null;
  toolName?: string | null;
  awaitMode?: string | null;
  cancelPolicy?: string | null;
};

export type InterruptRequestResult = {
  requestId: string;
  accepted: boolean;
  interruptRequestedAt?: string | null;
  alreadyInterrupted: boolean;
  stalePreview: boolean;
  preview?: CascadeCancelPreview | null;
};

export type DerivedCancelCauseView = {
  cause: "userCancelled" | "interrupted" | "deadline" | "unknown";
  source:
    | "requestInterrupt"
    | "parentCascade"
    | "deadline"
    | "toolLifecycle"
    | "responseInterruptedAt"
    | "unresolved";
  confidence: "direct" | "derived";
  at?: string | null;
  evidence: string[];
};

// Command request shapes (mirror bridge/types/requests/operations.rs).

export type DesktopOperationsSnapshotRequest = {
  agentDid?: string | null;
  rootRequestId?: string | null;
  includeTerminal?: boolean;
};

export type DesktopListSubagentTreeRequest = {
  rootRequestId: string;
  agentDid?: string | null;
  includeTerminal?: boolean;
  maxDepth?: number;
};

export type DesktopPreviewInterruptCascadeRequest = {
  requestId: string;
  agentDid?: string | null;
  includeTerminal?: boolean;
};

export type DesktopInterruptRequestRequest = {
  requestId: string;
  cause: "userCancelled";
  cascade: boolean;
  expectedPreviewSignature?: string | null;
};

// MCP health panel (panel #278) — mirrors `MCPServiceHealthView` and
// `McpServiceProbeResult` in bridge/types/views/operations.rs. `status`
// is the precise `HealthState.toDefraDB` projection from
// Proofs/MCPHealth/State.lean so the panel can distinguish back-off
// (`evicted`) from in-flight retry (`reconnecting`) — the public
// three-state `HealthStatus` collapses both to `unreachable`.
export type MCPServiceHealthView = {
  serviceId: string;
  agentDid?: string | null;
  endpoint?: string | null;
  status?: "healthy" | "degraded" | "evicted" | "reconnecting" | string | null;
  failureCount?: number | null;
  kMax?: number | null;
  backoffUntil?: string | null;
  lastProbeAt?: string | null;
  lastSeen?: string | null;
  lastErrorClass?: string | null;
  lastErrorMessage?: string | null;
  updatedAt?: string | null;
};

export type McpServiceProbeResult = {
  serviceId: string;
  status: string;
  latencyMs: number;
  lastError?: string | null;
};

export type DesktopProbeMcpServiceRequest = {
  serviceId: string;
};

export type WorkspaceEntryView = {
  name: string;
  kind: "dir" | "file";
  size?: number | null;
};

export type WorkspaceListingView = {
  root: string;
  subpath: string;
  entries: WorkspaceEntryView[];
  truncated: boolean;
};

export type HeldToolCallView = {
  toolCallId: string;
  requestId: string | null;
  sessionId: string | null;
  agentDid: string | null;
  toolName: string | null;
  args: string | null;
  deadlineAt: string | null;
};

export type DesktopResolveHoldRequest = {
  agentDid: string;
  toolCallId: string;
  approve: boolean;
  reason?: string | null;
};

export type ResolveHoldResult = {
  approvalId: string;
  toolCallId: string;
  decision: string;
};
